[Fix for #1787] Add Extension Methods to IEnumerable<IOptionDescriptor<object>>

tonysneed opened the pull request:
Addresses Issue #1787 by adding the following extension methods to IEnumerable<IOptionDescriptor<object>>:
* InstanceOf
* InstanceOfOrDefault
* InstancesOf
* RemoveTypesOf

These enable easier access for configuring options. For example, to configure the JsonOutputFormatter to preserve references and handle cycles:
```csharp
var formatter = options.OutputFormatters.InstanceOf<JsonOutputFormatter>();
formatter.SerializerSettings.PreserveReferences = true;
```

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+        public static TInstance InstanceOf<TInstance>([NotNull] this 
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            var instance = descriptors.Single(d => d.OptionType == typeof(TInstance)).Instance;
```
this should throw if there is no instance of the matching type rather than returning null (because now you have two patterns for the return, exception if the type is not there, and null if the descriptor uses a type rather than instance)

tonysneed replied:
Not sure if I understand what the issue is here. The call to Single will throw if there is no instance of the matching type.  The method will not return null.

yishaigalatzer replied:
Yes that was it, thank you

tonysneed replied:
Fixed with commit b4b2532e.

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+        public static IEnumerable<TInstance> InstancesOf<TInstance>([NotNull] this 
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            var instances = descriptors
```
similarly here, you are not guaranteed to have an instance, need to filter the null ones out.

tonysneed replied:
I'll refactor, taking into consideration that the Instance property of the descriptor could be null.

tonysneed replied:
Fixed with commit b4b2532e.

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <typeparam name="TOptionDescriptor">A type that implements 
+        /// <see cref="IOptionDescriptor{object}"/>.</typeparam>
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        public static void RemoveTypesOf<TInstance, TOptionDescriptor>([NotNull] this 
```
This is pretty ugly to use, we should either go with your other suggestion (do explicit extension methods for removes) or base it on IEnumerable and cast to IList if possible (didn't try this). @rynowak please share thoughts here

tonysneed replied:
Casting from IEnumerable to IList won't work.  That leaves explicit extension methods for `RemoveTypesOf`, and possibly `ReplaceTypesOf`, on `IList<InputFormatterDescriptor>` and `IList<OutputFormatterDescriptor>`.  Would you like me to do that?

Just to be sure, `RemoveTypesOf` should also remove descriptors with null Instances?  That's the way it's written now.

tonysneed commented:
So far I only modified `OutputFormatterDescriptor` and `OutputFormatterDescriptor` to implement `IOptionDescriptor<TOption>`.  Should I update all the other OptionDescriptors to do the same?

yishaigalatzer commented:
@tonysneed thanks for bringing it up, I believe you meant you changed `InputFormatterDescriptor` and `OutputFormatterDescriptor`. I can take a look at the others as well and get back to you, I think it makes sense to make the similar but I want to take some time and think it through

tonysneed commented:
Yes, typo in that last comment. :) I believe we'll want InstanceXxx methods to work across all the option descriptors.

tonysneed commented:
Commit ac7c0b8: I took the liberty of implementing explicit extension methods for `RemoveTypesOf` on `IList<InputFormatterDescriptor>` and `IList<OutputFormatterDescriptor>`. Let me know if that's acceptable. :cold_sweat:

yishaigalatzer commented:
This looks fine, here is the list of all the other

ExcludeValidation - No, we don't encourage removing stuff from here
InputFormatter - ** No** , clear should do it. There is just the json input formatter there
ModelBinder - ** Yes**
ModelValidatorProvider - No, clear should do it for 99% of cases if you want to replace.
OutputFormatter - ** Yes** (done)
ValueProviderFactory - ** Yes**, although not super common, might be useful to remove Form support.
ViewEngine - No, one entry by default, clear just does it.
ViewLocationExpander - No, Empty by default

tonysneed commented:
Thanks - I'll add the RemoveTypesOf extension method on just the ones listed here with Yes.

tonysneed commented:
@yishaigalatzer In commit 3d5a684 I added / removed `RemoveTypesOf` on the option descriptor lists as you requested.

Would you like me to update all the option descriptors to implement `IOptionDescriptor<TOption>`?

yishaigalatzer commented:
@tonysneed  Not a bad idea, lets do it and call it a day for this PR.

yishaigalatzer commented:
Also looks like something is messed up with the changes (now showing 313 files with lots of other peoples commits). You might want to rebase on dev again.

tonysneed commented:
Ah yes, I see what happened there.  After rebasing on dev, I ended up with commits 34aefb1b through ac7c0b8d listed as incoming:
```
On branch option-descriptor-extensions
Your branch is behind 'origin/option-descriptor-extensions' by 7 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
```
When I pulled in those commits, they duplicated the ones that had been rebased.  So I removed the rebased commits and pulled in the old ones, which also brought in the older commits from 8e91c1a onward.

I've been able to rebase locally on dev again, and like before I have the old incoming commits and git status gives me the same message.  If I try to push to origin, I get the following message:
```
To https://github.com/tonysneed/Mvc.git
 ! [rejected]        option-descriptor-extensions -> option-descriptor-extensions (non-fast-forward)
error: failed to push some refs to 'https://github.com/tonysneed/Mvc.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
The problem with pulling in the original commits, is that they duplicate the ones that have been rebased.  Instead, according to the [GitHub help on rebase] (https://help.github.com/articles/using-git-rebase/), I'll need to **force push**.  The reason I hesitate to do that is because it might make reviewing more difficult.  But seeing as we're getting toward the tail end of this PR, we should probably be OK.

So if it's alight, I can force push the rebase for this branch. Is that OK?  Hopefully that will solve the problem with my last attempt to rebase. :scream:

tonysneed commented:
FYI, here is what my history looks like after rebasing on dev.  As you can see, the commits that had been in the branch have been corrected rebased on dev.  But they are now also listed as incoming.

![git-rebase-incoming](https://cloud.githubusercontent.com/assets/2836367/6037812/c34959a0-ac57-11e4-9d4c-efb3b9c41865.png)

rynowak commented:
> So if it's alight, I can force push the rebase for this branch. Is that OK? Hopefully that will solve the problem with my last attempt to rebase. :scream:

Yes. force push to your own fork or your own branch is totally OK. 

Also while you're doing a rebase you can do `git rebase --skip` for a commit that's been subsumed by a change in another branch. 

I do this frequently if I'm working on feature `B` which depends on feature `A`. When I get CR feedback for `A`, I updated it and it eventually merge it to `dev` as `A'`. I then rebase `B` on top of dev... when I get merge conflicts trying to apply `A`, I do `git rebase --skip`.

You can also do this with `git rebase -i <branch>` by deleting lines

tonysneed commented:
I did a force push, and it totally worked!  The incoming commits are gone and all I have left is a nice clean history with my commits rebased on top of dev. I think I'm starting to get the hang of the PR workflow, and how to bring the branch up-to-date with rebase.

@rynowak Great tip on `git rebase --skip` for commits that have been superseded by what's already been merged. Especially like the hint re: deleting lines in an interactive rebase. :+1:

tonysneed commented:
@yishaigalatzer Is there anything else that needs to be done with this PR? Hopefully, the history is now clean and the all items have been addressed. :pray:

yishaigalatzer commented:
Squash and @dougbu s going to do a final pass then merge

tonysneed commented:
@yishaigalatzer @dougbu All squashed and ready for the final pass. Cheers! :beers:

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/IOptionDescriptor.cs:
```diff
@@ -0,0 +1,26 @@
+namespace Microsoft.AspNet.Mvc.OptionDescriptors
+{
+    /// <summary>
+    /// Encapsulates information that describes a <typeparamref name="TOption"/> option on <see cref="MvcOptions"/> .
```
teensy nit: remove space before the period

tonysneed replied:
Fixed in commit ec0fc095.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/IOptionDescriptor.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        /// <summary>
+        /// Gets the type of the <typeparamref name="TOption"/> described by this
+        /// <see cref="OptionDescriptor{TOption}"/>.
```
`OptionDescriptor` -> `IOptionDescriptor`

tonysneed replied:
Fixed in commit ec0fc095.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/InputFormatterDescriptor.cs:
```diff
@@ -13,7 +13,7 @@ public class InputFormatterDescriptor : OptionDescriptor<IInputFormatter>
         /// Creates a new instance of <see cref="InputFormatterDescriptor"/>.
         /// </summary>
-        /// <param name="type">A <see cref="IOutputFormatter"/> type that the descriptor represents.
+        /// <param name="type">A <see cref="IInputFormatter"/> type that the descriptor represents.
```
nice catch

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
```diff
@@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
+        /// Removes all instances of <typeparamref name="TInstance"/> from a list of ModelBinderDescriptors.
+        /// </summary>
+        /// <typeparam name="TInstance">A type that represents an <see cref="IModelBinder"/>.</typeparam>
+        /// <param name="descriptors">A list of ModelBinderDescriptors.</param>
```
nit: `<see cref="ModelBinderDescriptors"/>s`

dougbu replied:
and at line 86

tonysneed replied:
Line 89 is consistent with the xml comment for the descriptors parameter in the other methods in `ModelBinderDescriptorExtensions`.  And also with methods in all the other Xxx`DescriptorExtensions`.  Probably best not to change it just here. :grimacing:

tonysneed replied:
At line 86, to be consistent with other methods, 'list of ModelBinderDescriptors' should be changed to 'descriptor collection'?

dougbu replied:
hmm, I see your point.  the XML comments of this and similar extension classes are a bit inconsistent though "descriptor collection" seems most common. my fave is `<paramref name="descriptors"/>` but you can go with "descriptor collection" given what other methods are doing.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
```diff
@@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
+        /// <summary>
+        /// Removes all instances of <typeparamref name="TInstance"/> from a list of ModelBinderDescriptors.
+        /// </summary>
+        /// <typeparam name="TInstance">A type that represents an <see cref="IModelBinder"/>.</typeparam>
```
nit: tend to avoid "represents".  so perhaps `The <see cref="ModelBinderDescriptor"/> sub-type of instances to remove.`  not positive `<see cref="ModelBinderDescriptor"/> sub-` is that helpful.  up to you.

tonysneed replied:
I would say a "type that implements" `<see cref="IModelBinder"/>`.

dougbu replied:
sure

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
```diff
@@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
+        {
+            for (int i = descriptors.Count - 1; i >= 0; i--)
+            {
+                if (descriptors[i].OptionType == typeof(TInstance))
```
not positive of the expected behaviour here.  should this use `IsAssignableFrom()`?

tonysneed replied:
The behavior is as expected, because OptionType is set in both ctor's of OptionDescriptor, whether or not an Instance is assigned.  In other words, we're not really concerned with the Instance and are removing based just on OptionType.

dougbu replied:
if anyone sub-classes `BinderTypeBasedModelBinder` and calls `RemoveTypesOf<BinderTypeBasedModelBinder>`, the sub-classed instances will not be removed because those `OptionDescriptor<TOption>`s will have `OptionType=={that subclass}` even if created with an explicit `TOption`.  `OptionType` is very concerned about the instance's type when `Instance` is non-`null`.

probably not an issue as long as we don't expect / want anyone to create complicated `IModelBinder` (really `TOption` in general) type hierarchies. but worth `<remarks/>` in the XML comments about `exactly <typeparamref name="TInstance"/>`.

yishaigalatzer replied:
No this should be specific

-----Original Message-----
From: "Doug Bunting" <notifications@github.com>
Sent: ‎2/‎7/‎2015 11:40 AM
To: "aspnet/Mvc" <Mvc@noreply.github.com>
Cc: "Yishai Galatzer" <yishai_galatzer@yahoo.com>
Subject: Re: [Mvc] [Fix for #1787] Add Extension Methods toIEnumerable<IOptionDescriptor<object>> (#1906)

In src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
> @@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
>              descriptors.Insert(index, descriptor);
>              return descriptor;
>          }
> +
> +
> +        /// <summary>
> +        /// Removes all instances of <typeparamref name="TInstance"/> from a list of ModelBinderDescriptors.
> +        /// </summary>
> +        /// <typeparam name="TInstance">A type that represents an <see cref="IModelBinder"/>.</typeparam>
> +        /// <param name="descriptors">A list of ModelBinderDescriptors.</param>
> +        public static void RemoveTypesOf<TInstance>(
> +            [NotNull] this IList<ModelBinderDescriptor> descriptors)
> +            where TInstance : class, IModelBinder
> +        {
> +            for (int i = descriptors.Count - 1; i >= 0; i--)
> +            {
> +                if (descriptors[i].OptionType == typeof(TInstance))
not positive of the expected behaviour here. should this use IsAssignableFrom()?
—
Reply to this email directly or view it on GitHub.

dougbu replied:
@yishaigalatzer :ok_hand:

do you agree with `<remarks/>` to make this clear?

tonysneed replied:
@yishaigalatzer Let me know if you think I should add `<remarks/>` and what it should say.

dougbu replied:
@tonysneed if we need `<remarks/>`, we'll add them later.  don't block on @yishaigalatzer answering here.

yishaigalatzer replied:
I don't see the issue. You can only create something with a specific instance type, there is no `Create<T>`. If this should be documented it should go in the summary something along the lines of "removes options where the type exactly matches ...".

Again like everything else, these methods are for overriding default, and are convenience only, the user can always write his own (or we can add more later).

tonysneed replied:
Updated summary comment in commit 035e874b to include `where the type exactly matches <typeparamref name="TInstance"/>.`

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
```diff
@@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
+            {
+                if (descriptors[i].OptionType == typeof(TInstance))
+                {
+                    descriptors.Remove(descriptors[i]);
```
nit: `RemoveAt(i)` since `Remove()` first figures out the (known) index

tonysneed replied:
Fixed in commit f288931d. Also fixed `OutputFormatterDescriptorExtensions` and `ValueProviderFactoryDescriptorExtensions`.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/IOptionDescriptor.cs:
```diff
@@ -0,0 +1,26 @@
+    /// Encapsulates information that describes a <typeparamref name="TOption"/> option on <see cref="MvcOptions"/> .
+    /// </summary>
+    /// <typeparam name="TOption">The type of the option.</typeparam>
+    public interface IOptionDescriptor<out TOption>
```
to confirm: this interface exists only because we prefer not to extend the `OptionDescriptor<TOption>` class?  is it buying us anything else?

tonysneed replied:
The interface is needed for covariance, so that methods in OptionDescriptorExtensions extending `IEnumerable<IOptionDescriptor<object>>` can be called from a sequence of `InputFormatterDescriptor`, for example.

dougbu replied:
thanx

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <see cref="IOptionDescriptor{object}"/>; 
+        /// throws an exception if there is not exactly one <typeparamref name="TInstance"/> in the sequence.</returns>
+        public static TInstance InstanceOf<TInstance>([NotNull] this
+            IEnumerable<IOptionDescriptor<object>> descriptors)
```
nit: don't break this parameter over multiple lines

tonysneed replied:
Fixed in commit 74bba0dc.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        /// <returns>The only instance of <typeparamref name="TInstance"/> from a sequence of 
+        /// <see cref="IOptionDescriptor{object}"/>; 
+        /// throws an exception if there is not exactly one <typeparamref name="TInstance"/> in the sequence.</returns>
```
probably better to move this to
``` c#
<exception cref="InvalidOperationException">
Thrown if there is not exactly one <typeparamref name="TInstance"/> in the sequence.
</exception>
```

tonysneed replied:
Fixed in commit 74bba0dc. Fixed in both `InstanceOf` and `InstanceOfOrDefault`.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <returns>The only instance of <typeparamref name="TInstance"/> from a sequence of 
+        /// <see cref="IOptionDescriptor{object}"/>.</returns>
+        /// <exception cref="InvalidOperationException"> 
+        /// Thrown if there is not exactly one<typeparamref name="TInstance"/> in the sequence.</exception>
```
space after "one" here and on line 41

tonysneed replied:
Fixed in commit 21a6d4ba.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+    public static class OptionDescriptorExtensions
+    {
+        /// <summary>
+        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors.
```
@tonysneed @yishaigalatzer I find this method's behaviour surprising:
- expected a method with this name to `null` if none are found. put another way, why does `InstanceOfOrDefault()` exist?
- didn't expect this method (or `InstanceOfOrDefault()` if we need it) to `throw` when multiple matches are found. glad the behaviour is described but why aren't `InstanceOf()` and `InstancesOf()` more like for example `GetCustomAttribute()` and `GetCustomAttributes()`?
 - if we do want this behaviour, why defer to `Single()` to `throw` the `InvalidOperationException` rather than craft a specific message in MVC?

- expected this method to return any subclass of `TInstance` i.e. behave more like `OfType<TInstance>().FirstOrDefault()` (or `...Single()` if we stick w/ current behaviour on the other two points). but we're having that conversation elsewhere :smiley_cat:

yishaigalatzer replied:
InstanceOf() is supposed to throw, and InstanceOfOrDefault is that one that returns null

dougbu replied:
> `InstanceOf()` is supposed to `throw`, and `InstanceOfOrDefault` is that one that returns `null`

@yishaigalatzer why do we need both?

also both methods `throw` when multiple elements in the list are of the given type. is that desired?

generally, my expectation was for `...InstanceOf<TInstance>()` to be shorthand for `...Select(descriptor => descriptor.Instance).OfType<TInstance>().FirstOrDefault()`; ending with `...First()` if `throw`ing a system-provided `InvalidOperationException` was preferred.

dougbu replied:
well, `...Where(descriptor => descriptor.OptionType == typeof(TInstance) && descriptor.Instance != null).Select(descriptor => descriptor.Instance).FirstOrDefault()` given type-specificity we want (see [earlier](https://github.com/aspnet/Mvc/pull/1906#discussion_r24292104) discussion).  probably a `Cast()` in there somewhere too.

yishaigalatzer replied:
Because that's the design we asked him to implement. This is a code review and not a design review. 

Design wise:
The pattern follows normal .net API design.

Most users would use the overload that throw. Because it works as expected most of the time.

3rd party library developers will use the or default overload because they cannot make assumptions, or they will write their own method.

-----Original Message-----
From: "Doug Bunting" <notifications@github.com>
Sent: ‎2/‎7/‎2015 4:32 PM
To: "aspnet/Mvc" <Mvc@noreply.github.com>
Cc: "Yishai Galatzer" <yishai_galatzer@yahoo.com>
Subject: Re: [Mvc] [Fix for #1787] Add Extension Methods toIEnumerable<IOptionDescriptor<object>> (#1906)

In src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
> +﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
> +// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
> +
> +using System.Linq;
> +using System.Collections.Generic;
> +using Microsoft.AspNet.Mvc.OptionDescriptors;
> +
> +namespace Microsoft.AspNet.Mvc
> +{
> +    /// <summary>
> +    /// Extension methods for collections of option descriptors.
> +    /// </summary>
> +    public static class OptionDescriptorExtensions
> +    {
> +        /// <summary>
> +        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors.
InstanceOf() is supposed to throw, and InstanceOfOrDefault is that one that returns null
@yishaigalatzer why do we need both?
also both methods throw when multiple elements in the list are of the given type. is that desired?
generally, my expectation was for ...InstanceOf<TInstance>() to be shorthand for ...Select(descriptor => descriptor.Instance).OfType<TInstance>().FirstOrDefault(); ending with ...First() if throwing a system-provided InvalidOperationException was preferred.
—
Reply to this email directly or view it on GitHub.

tonysneed replied:
Fixed with commit 21a6d4ba: copy/paste error from prior commit. Exception for `InstanceOfOrDefault` now states more than one `<typeparamref name="TInstance"/>`.  This mirrors the verbiage for `Single` and `SingleOrDefault` in Linq;

dougbu replied:
@tonysneed thanks.  that's fine.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <typeparam name="TInstance">The type of the instance described by the option descriptor.</typeparam>
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        /// <returns>An IEnumerable of <typeparamref name="TInstance"/> that contains instances from a sequence 
+        /// of option descriptors.</returns>
```
not any random "sequence of option descriptors".  instead `... from <paramref name="descriptors"/>.`

tonysneed replied:
Changed to: sequence of `<see cref="IOptionDescriptor{object}"/>` to match other methods.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <summary>
+        /// Returns all the instances of <typeparamref name="TInstance"/> from a sequence of option descriptors.
+        /// </summary>
+        /// <typeparam name="TInstance">The type of the instance described by the option descriptor.</typeparam>
```
nit: "The type of the instances to find." (plural)

tonysneed replied:
Fixed in commit 21a6d4ba. Changed TInstance param in other methods to match.

dougbu commented on test/Microsoft.AspNet.Mvc.Core.Test/OptionDescriptors/InputFormatterDescriptorExtensionTest.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
```
why?

tonysneed replied:
Left over from `RemoveTypesOf` test which I removed after deleting the method from `InputFormatterDescriptorExtensions`. Will remove.

tonysneed replied:
Fixed in commit f5b229b7.

dougbu commented:
:watch: will look at tests after I understand the intentions here a bit more

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/ModelBinderDescriptorExtensions.cs:
```diff
@@ -80,5 +80,24 @@ public static class ModelBinderDescriptorExtensions
             descriptors.Insert(index, descriptor);
             return descriptor;
         }
+
```
nit: remove extra blank line

tonysneed replied:
Fixed in commit 035e874b.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+
+        /// <summary>
+        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors, 
+        /// or a default value if the sequence is empty.
```
nit: "if the sequence is empty" -> "if such an instance is not found"

tonysneed replied:
Think I'd like to keep as is, which matches description of [Enumerable.SingleOrDefault] (https://msdn.microsoft.com/en-us/library/vstudio/bb342451%28v=vs.100%29.aspx).

dougbu replied:
huh?  the relevant documentation is for `SingleOrDefault<TSource>(this IEnumerable<TSource>, Func<TSource, Boolean>)` (with my highlighting):
> Returns the only element of a sequence that satisfies a specified condition or a default value __if no such element exists__; this method throws an exception if more than one element satisfies the condition.

tonysneed replied:
You're looking at the overload of `SingleOrDefault` that takes a predicate. But the [doc] (https://msdn.microsoft.com/en-us/library/vstudio/bb342451(v=vs.110).aspx) for `SingleOrDefault<TSource>(this IEnumerable<TSource> source)` reads:
> Returns the only element of a sequence, or a default value **if the sequence is empty**; this method throws an exception if there is more than one element in the sequence.

Seems like `InstanceOfOrDefault` aligns with this version?

dougbu replied:
you're calling the other `SingleOrDefault()` overload -- the one that takes a predicate.  so, that's the one you've aligned with.  but this nit is not worth another round of changes.

tonysneed replied:
I see where you're coming from.  I was aligning with the signature of `InstanceOrDefault`.

tonysneed replied:
Sorry I didn't get this right away. The method is really filtering by type, but the single parameter threw me off. Glad to fix, if you're willing to hang on for another rebase. I'm in Europe, so it's a bit late here.

tonysneed replied:
Oops, just saw the merge come thru. :)

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        public static TInstance InstanceOf<TInstance>([NotNull] this IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            var instance = descriptors.Single(d => d.OptionType == typeof(TInstance)
+                && d.Instance != null).Instance;
```
teensy nit: would be slightly more readable if the condition were kept together.  e.g.
``` c#
var instance = descriptors
    .Single(d => d.OptionType == typeof(TInstance) && d.Instance != null)
    .Instance;
```
or all on one line if it fits.

dougbu replied:
nit: `descriptor` rather than `d` throughout this class.  then pick another fully-expanded name for `descriptor` variable in `InstanceOfOrDefault()`

tonysneed replied:
I don't think there's a hard and fast rule about using single letter variable names for lambda parameters.  I notice it's about half and half in Microsoft.AspNet.Mvc.Core. Inclined to leave it as is for now, unless @yishaigalatzer thinks otherwise.

tonysneed replied:
You won me over ... I renamed `d` ro `descriptor` in commit 9efc0929.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        {
+            TInstance instance = default(TInstance);
+            var descriptor = descriptors.SingleOrDefault(d => d.OptionType == typeof(TInstance)
+                && d.Instance != null);
```
similar teensy nit about keeping the condition together

tonysneed replied:
Fixed in commit 035e874b.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+        public static TInstance InstanceOfOrDefault<TInstance>([NotNull] this
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            TInstance instance = default(TInstance);
```
nit: move immediately above line 48 and use `var` if the compiler lets you

tonysneed replied:
Fixed in commit 035e874b.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
```diff
@@ -0,0 +1,71 @@
+            var instances = descriptors
+                .Where(d => d.OptionType == typeof(TInstance) && d.Instance != null)
+                .Select(d => d.Instance);
+            return instances.Cast<TInstance>();
```
nit: use `.Select(descriptor => (TInstance)descriptor.Instance)` to avoid additional Linq step

tonysneed replied:
Fixed in commit 035e874b.

dougbu commented on test/Microsoft.AspNet.Mvc.Core.Test/OptionDescriptors/OptionDescriptorExtensionsTest.cs:
```diff
@@ -0,0 +1,276 @@
+        }
+
+        [Fact]
+        public void InputFormatters_InstanceOf_ReturnsInstanceOfIInputFormatter()
```
nit: like the test names that include `IfThisCase` e.g. `IfOneExists` here

tonysneed replied:
Fixed in commit f5b229b7.

dougbu commented on test/Microsoft.AspNet.Mvc.Core.Test/OptionDescriptors/OptionDescriptorExtensionsTest.cs:
```diff
@@ -0,0 +1,276 @@
+
+namespace Microsoft.AspNet.Mvc.OptionDescriptors
+{
+    public class OptionDescriptorExtensionsTest
```
FYI one pattern for testing base classes that might have worked here is
``` c#
public abstract class OptionDescriptorExtensionsTestBase<TOption>
{
  protected abstract List<IOptionDescriptor<TOption>> GetDescriptors(MvcOptions options);
  protected abstract TOption GetInstance(); // return new instance on each call
  protected abstract TOption GetMockInstance(); // return new mock instance on each call
  ...
}
...
public class ExcludeValidationDescriptorTests : OptionDescriptorExtensionsTestBase<IExcludeTypeValidationFilter>
{
  ...
}
```
but I suspect in this case it would get fairly complex and covariance might be a blocker.

tonysneed replied:
Yes, these could use some refactoring. Good tip, but probably a bit much at this stage.

dougbu commented on test/Microsoft.AspNet.Mvc.Core.Test/OptionDescriptors/ModelBinderDescriptorExtensionsTest.cs:
```diff
@@ -68,5 +69,22 @@ public void ModelBinderDescriptors_AddsTypesAndInstances()
+            modelBinders.RemoveTypesOf<ByteArrayModelBinder>();
+
+            // Assert
+            Assert.False(modelBinders.Any(f => f.OptionType == typeof(ByteArrayModelBinder)));
```
nits:
- message when `Assert.False()` fails is basically useless so we tend to avoid it.  better to use `Assert.DoesNotContain(modelBinders, descriptor => descriptor.OptionType == typeof(ByteArrayModelBinder));`
- we also avoid abbreviations even for lambda parameters and even in test code

tonysneed replied:
Fixed in commit f5b229b7.

dougbu commented on test/XmlDataContractSerializerInputFormatterTests.cs:
```diff
@@ -1,499 +0,0 @@
-// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
just curious: these files show up as deletes because you haven't rebased recently?

tonysneed replied:
I haven't rebased recently from dev, so that's a possibility.  This file was renamed to XmlDataContractSerializerInputFormatterTest (no s at the end) in this branch.  I can rebase onto upstream before I rebase / squash this branch again.

dougbu commented:
:shipit: let me know when you've rebased / squashed branch (again, sorry) and removed the unnecessary change from from the commit.  up to you how many of the few remaining nits you address.  I'll push after you're done.

tonysneed commented:
@dougbu Note that in commit f5b229b7 I removed `#if ASPNET50` from OutputFormatterDescriptorExtensionTest and reordered some using directives.

tonysneed commented:
I squashed my commits and rebased onto upstream/dev, resolving the conflicts with `XmlDataContractSerializerInputFormatterTest` and `XmlDataContractSerializerOutputFormatterTest`. Should be ready to merge now. :pray: :smiling_imp:

dougbu commented:
Fixed with commits 2c21c7c9c7e83a058eb5aab7364119b52d2bc833 and 45aae5956d83006fdcba29cd464bb63d99323944.  The second avoided a build break.

@tonysneed thanks very much.  You can delete your branch now :smile_cat:

tonysneed commented:
Cool, and thanks to @dougbu for a thorough review! Thanks also to @yishaigalatzer & @rynowak. Glad to have contributed, and I learned a lot from the experience, especially about Git rebase and working with others on a pull request workflow. It will make a good topic for a future [blog] (http://blog.tonysneed.com) post. :smile: I'm looking forward to contributing more in the future!
